File: app/src/main/java/com/mxt/anitrend/presenter/widget/WidgetPresenter.java
package com.mxt.anitrend.presenter.widget;

import android.content.Context;
import android.os.AsyncTask;

import com.mxt.anitrend.base.custom.async.RequestHandler;
import com.mxt.anitrend.base.interfaces.event.RetroCallback;
import com.mxt.anitrend.presenter.base.BasePresenter;
import com.mxt.anitrend.util.KeyUtil;

import java.util.Locale;

/**
 * Created by max on 2017/10/31.
 */

public class WidgetPresenter<T> extends BasePresenter {

    private RequestHandler<T> mLoader;

    public static final int CONTENT_STATE = 0, LOADING_STATE = 1;

    public WidgetPresenter(Context context) {
        super(context);
    }

    /**
     * Template to make requests for various data types from api, the
     * <br/>
     * @param request_type the type of request to execute
     */
    public void requestData(@KeyUtil.RequestType int request_type, Context context, RetroCallback<T> callback) {
        mLoader = new RequestHandler<>(getParams(), callback, request_type);
        mLoader.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, context);
    }

    /**
     * Destroy any reference which maybe attached to
     * our context
     */
    @Override
    public void onDestroy() {
        if(mLoader != null && mLoader.getStatus() != AsyncTask.Status.FINISHED) {
            mLoader.cancel(true);
            mLoader = null;
        }
        super.onDestroy();
    }

    public static String convertToText(int count) {
        return String.format(Locale.getDefault()," %d ", count);
    }

    public static String valueFormatter(int size){
        if(size != 0){
            return size > 1000? String.format(Locale.getDefault(),"%.1f K", (float)size/1000): String.valueOf(size);
        }
        return "0";
    }
}

File: app/src/main/java/com/mxt/anitrend/presenter/widget/WidgetPresenter.kt
package com.mxt.anitrend.presenter.widget

import android.content.Context
import com.mxt.anitrend.base.custom.async.RequestHandler
import com.mxt.anitrend.base.interfaces.dao.BoxQuery
import com.mxt.anitrend.base.interfaces.event.RetroCallback
import com.mxt.anitrend.presenter.base.BasePresenter
import com.mxt.anitrend.util.KeyUtil
import com.mxt.anitrend.util.Settings
import java.util.Locale

/**
 * Created by max on 2017/10/31.
 */
class WidgetPresenter<T>(
    context: Context,
    boxQuery: BoxQuery,
    settings: Settings,
) : BasePresenter(context, boxQuery, settings) {
    private var loader: RequestHandler<T>? = null

    /**
     * Template to make requests for various data types from api, the
     *
     * @param requestType the type of request to execute
     */
    fun requestData(
        @KeyUtil.RequestType requestType: Int,
        context: Context,
        callback: RetroCallback<T>,
    ) {
        loader =
            RequestHandler(params, callback, requestType).also {
                it.execute(context)
            }
    }

    /**
     * Destroy any reference which maybe attached to
     * our context
     */
    override fun onDestroy() {
        loader?.cancel()
        loader = null
        super.onDestroy()
    }

    companion object {
        const val CONTENT_STATE = 0
        const val LOADING_STATE = 1

        fun convertToText(count: Int): String = String.format(Locale.getDefault(), " %d ", count)

        fun valueFormatter(size: Int): String {
            if (size != 0) {
                return if (size > 1000) {
                    String.format(Locale.getDefault(), "%.1f K", size / 1000f)
                } else {
                    size.toString()
                }
            }
            return "0"
        }
    }
}
